HTTP: Allow passing http client

SimonRichardson opened the pull request:
Passing in the http client, allows us to ensure the same client is used
throughout the lifecycle of a goose query. The previous implementations
either used the default http client in the stdlib or it created a new
http.client depending on circumstances of how you create the goose http
client.

The code breaks compatibility by providing a v3 of this library.

SimonRichardson committed:
HTTP: Allow passing http client

Passing in the http client, allows us to ensure the same client is used
throughout the lifecycle of a goose query. The previous implementations
either used the default http client in the stdlib or it created a new
http.client depending on circumstances of how you create the goose http
client.

The code in question tries to be as backwards compatible as possible so
we don't need to rev the library. There are some deprecated ways to
create a goose http client that are no longer in use internally, but are
still made available.

hmlanigan commented on http/client.go:
```diff
@@ -144,53 +152,47 @@ func New(options ...Option) *Client {
+	if client.Transport == nil {
+		client.Transport = &http.Transport{}
 	}
+	if t, ok := client.Transport.(*http.Transport); ok {
```
We should return an error if ok false. Otherwise you'd get a client without a TLS config and not know it.

hmlanigan commented on client/client.go:
```diff
@@ -226,7 +265,19 @@ func NewClientTLSConfig(creds *identity.Credentials, authMethod identity.AuthMod
+		client.Transport = &http.Transport{}
+	}
+	if t, ok := client.Transport.(*http.Transport); ok {
+		t.TLSClientConfig = config
```
We should return an error if ok false. Otherwise you'd get a client without a TLS config and not know it.

hmlanigan commented on http/client.go:
```diff
@@ -144,53 +152,47 @@ func New(options ...Option) *Client {
 
 // NewNonSSLValidating creates a new goose http *Client skipping SSL validation.
+// DEPRECATED (stickupkid): No longer used internally, left to ensure we don't
+// break compatibility.
```
Can we move to a v3, break compatibility and clean more up?  Especially with a few client methods needing to return an error.

SimonRichardson committed:
Client: Remove old deprecated methods

The http clients are now created in the client. We can remove the
various client functions in http package and now just pass options
correctly in client.

hmlanigan commented:
@SimonRichardson description of the PR needs updates related to v3

hmlanigan commented:
@SimonRichardson the description needs updating about v3.

hmlanigan reviewed (approved):
Great!  LGTM.
